src/System.Private.CoreLib/shared/System/Guid.cs

```diff
@@ -624,327 +628,108 @@ private static bool TryParseGuidWithHexPrefix(ReadOnlySpan<char> guidString, ref
                     numLen = guidString.Slice(numStart).IndexOf(',');
                     if (numLen <= 0)
                     {
-                        result.SetFailure(ParseFailureKind.Format, nameof(SR.Format_GuidComma));
+                        result.SetFailure(overflow: false, nameof(SR.Format_GuidComma));
                         return false;
                     }
                 }
-                else       // last case ends with '}', not ','
+                else // last case ends with '}', not ','
                 {
                     numLen = guidString.Slice(numStart).IndexOf('}');
                     if (numLen <= 0)
                     {
-                        result.SetFailure(ParseFailureKind.Format, nameof(SR.Format_GuidBraceAfterLastNumber));
+                        result.SetFailure(overflow: false, nameof(SR.Format_GuidBraceAfterLastNumber));
                         return false;
                     }
                 }
 
                 // Read in the number
-                int signedNumber;
-                if (!StringToInt(guidString.Slice(numStart, numLen), -1, ParseNumbers.IsTight, out signedNumber, ref result))
+                uint byteVal;
+                if (!TryParseHex(guidString.Slice(numStart, numLen), out byteVal, ref overflow) || overflow || byteVal > byte.MaxValue)
                 {
+                    // The previous implementation had some odd inconsistencies, which are carried forward here.
+                    // The byte values in the X format are treated as integers with regards to overflow, so
+                    // a "byte" value like 0xddd in Guid's ctor results in a FormatException but 0xddddddddd results
+                    // in OverflowException.
+                    result.SetFailure(overflow,
+                        overflow ? nameof(SR.Overflow_UInt32) :
+                        byteVal > byte.MaxValue ? nameof(SR.Overflow_Byte) :
+                        nameof(SR.Format_GuidInvalidChar));
                     return false;
                 }
-                uint number = (uint)signedNumber;
-
-                // check for overflow
-                if (number > 255)
-                {
-                    result.SetFailure(ParseFailureKind.Format, nameof(SR.Overflow_Byte));
-                    return false;
-                }
-                bytes[i] = (byte)number;
+                guidByteRef = (byte)byteVal;
```

Comment:
It should be a tiny bit faster/smaller to just do 
`Unsafe.Add(ref result._parsedGuid._d, i) = (byte)byteVal;` and not cache it in guidByteRef.

Reply:
Ok, will change. Thanks.